Add cross-dag dependencies guide

kentdanas opened the pull request:
@jwitz using one of my guide review coupons :)
Would love your thoughts on whether all the explanations are clear enough (especially in the last section), or whether more "something" (screenshots, words, ?) is needed. Didn't want to make it too repetitive, but also want to make sure everything is clear.

jwitz commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,247 @@
+
+There are a couple of things to note when using this operator:
+
+- If your dependent DAG requires a config input, or a specific execution date, these can be specified in the operator using the `conf` and `execution_date` params respectively.
```
```suggestion
- If your dependent DAG requires a config input or a specific execution date, these can be specified in the operator using the `conf` and `execution_date` params respectively.
```

jwitz commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,247 @@
+
+### ExternalTaskSensor
+
+The next method for creating cross-DAG dependencies is using the `ExternalTaskSensor`. This method uses a sensor in the child DAG, which will wait until a task is completed in the parent DAG before moving on to other tasks. You can find more info on this sensor [here](https://registry.astronomer.io/providers/apache-airflow/modules/externaltasksensor).
```
```suggestion
The next method for creating cross-DAG dependencies is to add a `ExternalTaskSensor` to your child DAG. The child DAG will wait until a task is completed in the parent DAG before moving on to other tasks. You can find more info on this sensor on the [Astronomer Registry](https://registry.astronomer.io/providers/apache-airflow/modules/externaltasksensor).
```

jwitz commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,247 @@
+
+The next method for creating cross-DAG dependencies is using the `ExternalTaskSensor`. This method uses a sensor in the child DAG, which will wait until a task is completed in the parent DAG before moving on to other tasks. You can find more info on this sensor [here](https://registry.astronomer.io/providers/apache-airflow/modules/externaltasksensor).
+
+This method is not as flexible as the `TriggerDagRunOperator`, since the dependency is implemented in the downstream DAG. It is ideal in situations where you have a DAG that can only run after another task or DAG(s) has completed. An example DAG using the `ExternalTaskSensor` is shown below:
```
```suggestion
This method is not as flexible as the `TriggerDagRunOperator`, since the dependency is implemented in the downstream DAG. It is ideal in situations where you have a child DAG that can run only after another task in a parent DAG has completed. An example DAG using the `ExternalTaskSensor` is shown below:
```
Rewriting this based on what I assume is going on here, but not super sure especially since we're no longer using the parent/child language.

jwitz reviewed (changes requested):
@kentdanas Started suggesting inline changes here, but I hit a wall with understanding what's going on at a high level with dependencies. Before continuing my review, I think this guide could benefit from a few inclusions:

* In the intro, we should define what "parent/child" is. Is it different from "upstream/downstream", and how so? Is it a dependency in and of itself, or is the dependency a separate relationship? Once you get that down, it'll be easier to explain each individual dependency scenario. 
* Even though it's only available in 2.1, could we show each of these methods as they appear in the DAG Dependencies view? We can mention at the beginning that the feature's only available in 2.1 and describe how to access it in the same place as we do now, but having that native Airflow visualization would be super helpful for understanding these examples. 

I'll take another look after you look through these current comments. Excited get this published!

kentdanas commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,247 @@
+
+When designing Airflow DAGs, it is often recommended to put all related tasks in the same DAG. However, sometimes this is not practical, and it is necessary to create dependencies between DAGs. According to the Airflow documentation on [cross-DAG dependencies](https://airflow.apache.org/docs/apache-airflow/stable/howto/operator/external_task_sensor.html#cross-dag-dependencies), this can be useful ...
+
+- Two DAGs are dependent but have different schedules
+- Different teams are responsible for the DAGs but there are cross-DAG dependencies
+- A task depends on a task in another DAG, but for a different execution date
```
I think it's worded poorly in the Apache doc - they're trying to say that even though the two tasks might be in the same DAG, one task depends on the other for a different execution date, which can't be accomplished in the same DAG Run. So essentially you would have a DAG that has a dependency with itself. That seemed unnecessarily confusing to me, so I simplified a bit here.

kentdanas commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,247 @@
+
+The `TriggerDagRunOperator` is an easy way to implement cross-DAG dependencies. This operator allows you to have a task in one DAG that triggers another DAG in the same Airflow environment. The Astronomer Registry has more documentation for this operator [here](https://registry.astronomer.io/providers/apache-airflow/modules/triggerdagrunoperator).
+
+The `TriggerDagRunOperator` is ideal in situations where you have one parent DAG, that needs to trigger one or more downstream child DAGs, or if you have dependent DAGs that have both upstream and downstream tasks in the parent DAG. Because you can use this operator for any task in your DAG, it is highly flexible; it can also often be an ideal replacement for SubDAGs. 
```
Went through and made the metaphors consistent, and also added some graph view screenshots for the example DAGs. Hopefully it's more clear now.

kentdanas commented:
@jwitz ok, I updated the language throughout to only use upstream/downstream (I liked this better than parent/child), and re-worked the intro to more thoroughly describe what those terms mean. Also added some clarifying language and screenshots of the example DAGs in the implementation section. Let me know if this helps make it clearer :)

jwitz commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,261 @@
+
+## Overview
+
+When designing Airflow DAGs, it is often best practice to put all related tasks in the same DAG. However, it's sometimes necessary to create dependencies between your DAGs. One way of thinking about this is that a node in your DAG is actually a whole other DAG, rather than just a single task. Throughout this guide, we'll use "upstream DAG" to denote the DAG (or partial DAG) that must run before t ...
```
Also...! do we need to more explicitly define "dependent DAG"? My impression is that both upstream and downstream DAGs are dependent DAGs, but it seems like it's mentioned only in reference to downstream DAGs.

kentdanas replied:
Both the "upstream" and the "downstream" DAG would be considered dependent DAGs (they are dependent on each other). This isn't any different than task dependencies, which are well known to any airflow user, so I don't really think it's necessary to define it explicitly.

jwitz commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,261 @@
+
+The next method for creating cross-DAG dependencies is to add a `ExternalTaskSensor` to your downstream DAG. The downstream DAG will wait until a task is completed in the upstream DAG before moving on to the rest of the DAG. You can find more info on this sensor on the [Astronomer Registry](https://registry.astronomer.io/providers/apache-airflow/modules/externaltasksensor).
+
+This method is not as flexible as the `TriggerDagRunOperator`, since the dependency is implemented in the downstream DAG. It is ideal in situations where you have a downstream DAG that can run only after another task in a upstream DAG has completed. An example DAG using the `ExternalTaskSensor` is shown below:
```
```suggestion
This method is not as flexible as the `TriggerDagRunOperator`, since the dependency is implemented in the downstream DAG. It is ideal in situations where you have a downstream DAG that can run only after a task in the upstream DAG has completed. An example DAG using the `ExternalTaskSensor` is shown below:
```

The "ideal situation" here seems to be reiterating the definition of the TaskSensor itself, is there anything else we can say about the context here that will give people a sense of whether to use this over the `TriggerDagRunOperator`?

kentdanas replied:
Yep, good call. Changed to an example that wouldn't work with the `TriggerDagRunOperator`.

jwitz commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,261 @@
+
+In this DAG, `downstream_task1` waits for the `bash_print_date2` task of `example_dag` to complete before moving on to execute the rest of the downstream tasks (`downstream_task2`). The graph view of the DAG looks like this:
+
+![External Task Sensor Grapg](https://assets2.astronomer.io/main/guides/cross-dag-dependencies/external_task_sensor_graph.png)
```
Names in this screenshot are inconsistent with the names in the sample DAG.

kentdanas replied:
Good catch, I updated the screenshot.

jwitz commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,261 @@
+
+![External Task Sensor Grapg](https://assets2.astronomer.io/main/guides/cross-dag-dependencies/external_task_sensor_graph.png)
+
+If you want the downstream DAG to wait for the entire upstream DAG to finish instead of a specific task, you can set the `external_task_id` to `None`. Note that in this case we specify that the external task must have a state of `success` in order for the downstream task to succeed, as defined by the `allowed_states` and `failed_states`. 
```
```suggestion
If you want the downstream DAG to wait for the entire upstream DAG to finish instead of a specific task, you can set the `external_task_id` to `None`. In this case, we specify that the external task must have a state of `success` for the downstream task to succeed, as defined by the `allowed_states` and `failed_states`. 
```

jwitz commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,261 @@
+
+### Airflow API
+
+The [Airflow API](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html) is another way of creating cross-DAG dependencies. This is especially useful in [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0), which has a full stable REST API. To use the API to trigger a DAG run, you can make a POST request to the `DAGRuns` endpoint as described in the Airfl ...
```
```suggestion
The [Airflow API](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html) is another way of creating cross-DAG dependencies. This is especially useful in [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0), which has a full stable REST API. To use the API to trigger a DAG run, you can make a POST request to the `DAGRuns` endpoint as described in the [Airflow documentation](https://www.astronomer.io/blog/introducing-airflow-2-0).
```

jwitz commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,261 @@
+
+The [Airflow API](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html) is another way of creating cross-DAG dependencies. This is especially useful in [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0), which has a full stable REST API. To use the API to trigger a DAG run, you can make a POST request to the `DAGRuns` endpoint as described in the Airfl ...
+
+This method is useful if your dependent DAGs live in different Airflow environments (more on this below), or if the downstream DAG does not have any downstream dependencies in the upstream DAG (e.g. the downstream DAG is the last task in the upstream DAG). One drawback to this method is that the task triggering the downstream DAG will complete once the API call is complete, rather than when the d ...
```
For my own knowledge, why is that last sentence a drawback?

kentdanas replied:
If you have a task A that has a downstream dependency (more tasks that run after it), you generally want to wait until that task A is completed before moving on the downstream tasks. With the API in this example, your task A is an entire DAG, and taks A will be considered complete once the API call is made, not when the DAG that got triggered finishes. So if your "downstream" DAG takes 3 minutes to complete, it will still be running even though task A is done, which may impact other downstream tasks.

jwitz commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,261 @@
+
+The [Airflow API](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html) is another way of creating cross-DAG dependencies. This is especially useful in [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0), which has a full stable REST API. To use the API to trigger a DAG run, you can make a POST request to the `DAGRuns` endpoint as described in the Airfl ...
+
+This method is useful if your dependent DAGs live in different Airflow environments (more on this below), or if the downstream DAG does not have any downstream dependencies in the upstream DAG (e.g. the downstream DAG is the last task in the upstream DAG). One drawback to this method is that the task triggering the downstream DAG will complete once the API call is complete, rather than when the d ...
```
```suggestion
This method is useful if your dependent DAGs live in different Airflow environments (more on this in the Cross-Deployment Dependencies section below), or if the downstream DAG does not have any downstream dependencies in the upstream DAG (e.g. the downstream DAG is the last task in the upstream DAG). One drawback to this method is that the task triggering the downstream DAG will complete once the API call is complete, rather than when the downstream DAG is complete.
```

jwitz commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,261 @@
+
+In order to use the `SimpleHttpOperator` to trigger another DAG, you need to define the following:
+
+- `endpoint`: this should be of the form `'/api/v1/dags/<dag-id>/dagRuns'` where `<dag-id>` is the ID of the DAG you want to trigger.
```
```suggestion
- `endpoint`: This should be of the form `'/api/v1/dags/<dag-id>/dagRuns'` where `<dag-id>` is the ID of the DAG you want to trigger.
```

jwitz commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,261 @@
+In order to use the `SimpleHttpOperator` to trigger another DAG, you need to define the following:
+
+- `endpoint`: this should be of the form `'/api/v1/dags/<dag-id>/dagRuns'` where `<dag-id>` is the ID of the DAG you want to trigger.
+- `data`: to trigger a DAG Run using this endpoint you must provide an execution date. In the example above, we use the `execution_date` of the upstream DAG, but this can be any date of your choosing. You can also specify other information about the DAG run as described in the API documentation linked above.
```
```suggestion
- `data`: To trigger a DAG Run using this endpoint, you must provide an execution date. In the example above, we use the `execution_date` of the upstream DAG, but this can be any date of your choosing. You can also specify other information about the DAG run as described in the API documentation linked above.
```

jwitz commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,261 @@
+
+- `endpoint`: this should be of the form `'/api/v1/dags/<dag-id>/dagRuns'` where `<dag-id>` is the ID of the DAG you want to trigger.
+- `data`: to trigger a DAG Run using this endpoint you must provide an execution date. In the example above, we use the `execution_date` of the upstream DAG, but this can be any date of your choosing. You can also specify other information about the DAG run as described in the API documentation linked above.
+- `http_conn_id`: this should be an Airflow connection of type HTTP, with your Airflow domain as the Host, and any authentication provided either as a Login/Password (if using Basic auth), or as a json-formatted Extra. In the example below we use an authorization token.
```
```suggestion
- `http_conn_id`: This should be an Airflow connection of type HTTP, with your Airflow domain as the Host. Any authentication should be provided either as a Login/Password (if using Basic auth) or as a JSON-formatted Extra. In the example below, we use an authorization token.
```

jwitz commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,261 @@
+
+## Cross-Deployment Dependencies
+
+Sometimes it may be necessary to implement cross-DAG dependencies where the DAGs do not exist in the same Airflow deployment. The `TriggerDagRunOperator` and `ExternalTaskSensor` methods described above are designed to work with DAGs in the same Airflow environment, so are not ideal for cross-Airflow deployments. The Airflow API on the other hand, is perfect for this use case. In this section we' ...
```
```suggestion
Sometimes it may be necessary to implement cross-DAG dependencies where the DAGs do not exist in the same Airflow deployment. The `TriggerDagRunOperator` and `ExternalTaskSensor` methods described above are designed to work with DAGs in the same Airflow environment, so they are not ideal for cross-Airflow deployments. The Airflow API, on the other hand, is perfect for this use case. In this section, we'll focus on how to implement this method on the Astronomer platform, but the general concepts will likely be similar wherever your Airflow environments are deployed.
```

jwitz reviewed (commented):
This is getting close @kentdanas ! Great job redefining some of the terms here. I left a few more comments and questions about the intro and the screen shots, but the rest of the guide looks great.

kentdanas commented:
@jwitz thanks again for your review! I think I addressed all your remaining comments, let me know if you see anything else!

jwitz commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,264 @@
+
+### ExternalTaskSensor
+
+The next method for creating cross-DAG dependencies is to add a `ExternalTaskSensor` to your downstream DAG. The downstream DAG will wait until a task is completed in the upstream DAG before moving on to the rest of the DAG. You can find more info on this sensor on the [Astronomer Registry](https://registry.astronomer.io/providers/apache-airflow/modules/externaltasksensor).
```
```suggestion
The next method for creating cross-DAG dependencies is to add an `ExternalTaskSensor` to your downstream DAG. The downstream DAG will wait until a task is completed in the upstream DAG before moving on to the rest of the DAG. You can find more info on this sensor on the [Astronomer Registry](https://registry.astronomer.io/providers/apache-airflow/modules/externaltasksensor).
```

jwitz commented on guides/cross-dag-dependencies.md:
```diff
@@ -0,0 +1,264 @@
+
+def print_task_type(**kwargs):
+    """
+    Dummy function to call before and after depdendent DAG.
```
```suggestion
    Dummy function to call before and after dependent DAG.
```

jwitz reviewed (approved):
Some minor fixes suggested, but otherwise looks good to go! 😎
